From 6952e5ac79a60b5f15d6dea0174b657b54a60044 Mon Sep 17 00:00:00 2001
From: Nikita Yushchenko <nikita.yoush@cogentembedded.com>
Date: Wed, 17 Aug 2022 13:09:06 +0300
Subject: [PATCH 59/92] pinctrl: renesas: r8a779f0: set MODSEL4 bits for
 MSPIn_SC

To get MSPI SC output, need to not only select pin function in IPSR, but
also set pin direction in MODSEL4.

Note: this only has support for SC output (spi master), not for SC input
(spi slave).

Signed-off-by: Nikita Yushchenko <nikita.yoush@cogentembedded.com>
---
 drivers/pinctrl/renesas/pfc-r8a779f0.c | 52 ++++++++++++++++++++++----
 1 file changed, 45 insertions(+), 7 deletions(-)

diff --git a/drivers/pinctrl/renesas/pfc-r8a779f0.c b/drivers/pinctrl/renesas/pfc-r8a779f0.c
index c0657223f900..cfc97ba29021 100644
--- a/drivers/pinctrl/renesas/pfc-r8a779f0.c
+++ b/drivers/pinctrl/renesas/pfc-r8a779f0.c
@@ -510,6 +510,14 @@ FM(IP0SR7_31_28)	IP0SR7_31_28	FM(IP1SR7_31_28)	IP1SR7_31_28	FM(IP2SR7_31_28)	IP2
 #define MOD_SEL1_2_3		FM(SEL_I2C1_0)	F_(0, 0)	F_(0, 0)	FM(SEL_I2C1_3)
 #define MOD_SEL1_0_1		FM(SEL_I2C0_0)	F_(0, 0)	F_(0, 0)	FM(SEL_I2C0_3)
 
+/* MOD_SEL4 */			/* 0 */			/* 1 */
+#define MOD_SEL4_5		FM(SEL_MSPI5_SC_IN)	FM(SEL_MSPI5_SC_OUT)
+#define MOD_SEL4_4		FM(SEL_MSPI4_SC_IN)	FM(SEL_MSPI4_SC_OUT)
+#define MOD_SEL4_3		FM(SEL_MSPI3_SC_IN)	FM(SEL_MSPI3_SC_OUT)
+#define MOD_SEL4_2		FM(SEL_MSPI2_SC_IN)	FM(SEL_MSPI2_SC_OUT)
+#define MOD_SEL4_1		FM(SEL_MSPI1_SC_IN)	FM(SEL_MSPI1_SC_OUT)
+#define MOD_SEL4_0		FM(SEL_MSPI0_SC_IN)	FM(SEL_MSPI0_SC_OUT)
+
 #define PINMUX_MOD_SELS \
 \
 MOD_SEL1_10_11 \
@@ -517,7 +525,13 @@ MOD_SEL1_8_9 \
 MOD_SEL1_6_7 \
 MOD_SEL1_4_5 \
 MOD_SEL1_2_3 \
-MOD_SEL1_0_1
+MOD_SEL1_0_1 \
+MOD_SEL4_5 \
+MOD_SEL4_4 \
+MOD_SEL4_3 \
+MOD_SEL4_2 \
+MOD_SEL4_1 \
+MOD_SEL4_0
 
 #define PINMUX_PHYS \
 	FM(SCL0) FM(SDA0) FM(SCL1) FM(SDA1) FM(SCL2) FM(SDA2) FM(SCL3) FM(SDA3) \
@@ -778,7 +792,7 @@ static const u16 pinmux_data[] = {
 
 	/* IP0SR4 */
 	PINMUX_IPSR_GPSR(IP0SR4_3_0,	GP4_00),
-	PINMUX_IPSR_GPSR(IP0SR4_3_0,	MSPI4SC),
+	PINMUX_IPSR_MSEL(IP0SR4_3_0,	MSPI4SC,	SEL_MSPI4_SC_OUT),
 	PINMUX_IPSR_GPSR(IP0SR4_3_0,	TAUD0I2),
 	PINMUX_IPSR_GPSR(IP0SR4_3_0,	TAUD0O2),
 
@@ -803,7 +817,7 @@ static const u16 pinmux_data[] = {
 	PINMUX_IPSR_GPSR(IP0SR4_19_16,	MSPI4SSI_N),
 	PINMUX_IPSR_GPSR(IP0SR4_19_16,	TAUD0I5),
 	PINMUX_IPSR_GPSR(IP0SR4_19_16,	TAUD0O5),
-	PINMUX_IPSR_GPSR(IP0SR4_19_16,	MSPI5SC),
+	PINMUX_IPSR_MSEL(IP0SR4_19_16,	MSPI5SC,	SEL_MSPI5_SC_OUT),
 
 	PINMUX_IPSR_GPSR(IP0SR4_23_20,	GP4_05),
 	PINMUX_IPSR_GPSR(IP0SR4_23_20,	MSPI4CSS3),
@@ -882,7 +896,7 @@ static const u16 pinmux_data[] = {
 	PINMUX_IPSR_GPSR(IP2SR4_15_12,	TAUD1I4),
 	PINMUX_IPSR_GPSR(IP2SR4_15_12,	TAUD1O4),
 
-	PINMUX_IPSR_GPSR(IP2SR4_19_16,	MSPI0SC),
+	PINMUX_IPSR_MSEL(IP2SR4_19_16,	MSPI0SC,	SEL_MSPI0_SC_OUT),
 	PINMUX_IPSR_GPSR(IP2SR4_19_16,	MSPI1CSS7),
 	PINMUX_IPSR_GPSR(IP2SR4_19_16,	TAUD1I5),
 	PINMUX_IPSR_GPSR(IP2SR4_19_16,	TAUD1O5),
@@ -921,7 +935,7 @@ static const u16 pinmux_data[] = {
 	PINMUX_IPSR_GPSR(IP3SR4_15_12,	TAUD1I13),
 	PINMUX_IPSR_GPSR(IP3SR4_15_12,	TAUD1O13),
 
-	PINMUX_IPSR_GPSR(IP3SR4_19_16,	MSPI1SC),
+	PINMUX_IPSR_MSEL(IP3SR4_19_16,	MSPI1SC,	SEL_MSPI1_SC_OUT),
 	PINMUX_IPSR_GPSR(IP3SR4_19_16,	MSPI0CSS2),
 	PINMUX_IPSR_GPSR(IP3SR4_19_16,	TAUD1I10),
 	PINMUX_IPSR_GPSR(IP3SR4_19_16,	TAUD1O10),
@@ -1027,7 +1041,7 @@ static const u16 pinmux_data[] = {
 	PINMUX_IPSR_GPSR(IP0SR7_7_4,	CAN0RX_INTP0),
 	PINMUX_IPSR_GPSR(IP0SR7_7_4,	RSENT0RX),
 	PINMUX_IPSR_GPSR(IP0SR7_7_4,	RSENT0RX_RSENT0SPCO),
-	PINMUX_IPSR_GPSR(IP0SR7_7_4,	MSPI2SC),
+	PINMUX_IPSR_MSEL(IP0SR7_7_4,	MSPI2SC,	SEL_MSPI2_SC_OUT),
 
 	PINMUX_IPSR_GPSR(IP0SR7_11_8,	CAN1TX),
 	PINMUX_IPSR_GPSR(IP0SR7_11_8,	RSENT1SPCO),
@@ -1082,7 +1096,7 @@ static const u16 pinmux_data[] = {
 	PINMUX_IPSR_GPSR(IP1SR7_23_20,	CAN6RX_INTP6),
 	PINMUX_IPSR_GPSR(IP1SR7_23_20,	RSENT6RX),
 	PINMUX_IPSR_GPSR(IP1SR7_23_20,	RSENT6RX_RSENT6SPCO),
-	PINMUX_IPSR_GPSR(IP1SR7_23_20,	MSPI3SC),
+	PINMUX_IPSR_MSEL(IP1SR7_23_20,	MSPI3SC,	SEL_MSPI3_SC_OUT),
 
 	PINMUX_IPSR_GPSR(IP1SR7_27_24,	CAN7TX),
 	PINMUX_IPSR_GPSR(IP1SR7_27_24,	RSENT7SPCO),
@@ -3716,6 +3730,30 @@ static const struct pinmux_cfg_reg pinmux_config_regs[] = {
 		MOD_SEL1_2_3
 		MOD_SEL1_0_1))
 	},
+	{ PINMUX_CFG_REG_VAR("MOD_SEL4", 0xdfd90100, 32,
+			     GROUP(4, 4, 4, 4, 4, 4, 2, 1, 1, 1, 1, 1, 1),
+			     GROUP(
+		/* RESERVED 31, 30, 29, 28 */
+		0, 0, 0, 0, 0, 0, 0, 0,	0, 0, 0, 0, 0, 0, 0, 0,
+		/* RESERVED 27, 26, 25, 24 */
+		0, 0, 0, 0, 0, 0, 0, 0,	0, 0, 0, 0, 0, 0, 0, 0,
+		/* RESERVED 23, 22, 21, 20 */
+		0, 0, 0, 0, 0, 0, 0, 0,	0, 0, 0, 0, 0, 0, 0, 0,
+		/* RESERVED 19, 18, 17, 16 */
+		0, 0, 0, 0, 0, 0, 0, 0,	0, 0, 0, 0, 0, 0, 0, 0,
+		/* RESERVED 15, 14, 13, 12 */
+		0, 0, 0, 0, 0, 0, 0, 0,	0, 0, 0, 0, 0, 0, 0, 0,
+		/* RESERVED 11, 10, 9, 8 */
+		0, 0, 0, 0, 0, 0, 0, 0,	0, 0, 0, 0, 0, 0, 0, 0,
+		/* RESERVED 7, 6 */
+		0, 0, 0, 0,
+		MOD_SEL4_5
+		MOD_SEL4_4
+		MOD_SEL4_3
+		MOD_SEL4_2
+		MOD_SEL4_1
+		MOD_SEL4_0))
+	},
 	{ },
 };
 
-- 
2.30.2

